### PR TITLE
docstrings added to iquick script

### PR DESCRIPTION
**Pull request by yoctoyotta1024**
Docstrings to aid understanding of script and relation to MIRA35 Manual.

Some docstrings are posisbly too verbose - should we reword them or put them in a documentation  e.g.  a readme instead?

PR also includes four minor changes not related to docstrings:
1) [ensure memmap of file is read only](https://github.com/halo-narval/hamp_radar/pull/2/commits/c9cc055fae5e79777b952df33f39c1959acfc8e3)
2) [refactor: explicit naming of kwarg keys](https://github.com/halo-narval/hamp_radar/pull/2/commits/12174c05178404791b3007a20f0ed366e9faa1ee)
3) [refactor: pass only bytes required by function for clarity](https://github.com/halo-narval/hamp_radar/pull/2/commits/65ef44f951c6ff05af0949280883d69c60a334a0)
4) [refactor: remove redundant list creation](https://github.com/halo-narval/hamp_radar/pull/2/commits/601a88b64c5fb12ddffb49dc50cc95c91e9e7101)

### CODE REVIEW

File: hamp_radar/iquick.py [hunk 1]
```diff
@@ -24,6 +55,25 @@ class SingleMainBlockGeometry:
 
 @dataclass
 class MultiMainBlockGeometry:
+    """
+    Representation of multiple main blocks as list of SingleSubBlockGeometry
+    instances with some overarching metadata. This can be multiple 'main chunks'
```
Comment (d70-t) [inline hamp_radar/iquick.py:60]:
```suggestion
    Representation of multiple consecutive main blocks with identical sub block geometries.
    This are multiple 'main chunks'
```

File: hamp_radar/iquick.py [hunk 2]
```diff
@@ -24,6 +55,25 @@ class SingleMainBlockGeometry:
...
+    as in Meteorological Ka-Band Cloud Radar MIRA35 Manual, section 2.3.2
+    'File structure'.
+    
+    Allows for 'count' number of SingleMainBlockGeometry instances to be
+    combined, assuming they are similar enough (e.g. same subblocks, same offset,
+    same tag etc.). (See compact_geometry function.)
```
Comment (d70-t) [inline hamp_radar/iquick.py:66]:
```suggestion
    Allows for 'count' number of main blocks (and their subblocks) to be read at once,
    assuming they are similar enough. (See compact_geometry function.)
```

File: hamp_radar/iquick.py [hunk 3]
```diff
@@ -24,6 +55,25 @@ class SingleMainBlockGeometry:
...
+
+    Attributes:
+        tag (str): The tag/signature describing the main blocks.
+        offset (int): Pointer to the start of the first main block.
+        count (int): The number of SingleMainBlockGeometry instances
+                     combined to make the MultiMainBlockGeometry instance.
```
Comment (d70-t) [inline hamp_radar/iquick.py:72]: It's not a combination of sub blocks, it's a combination of main blocks.
```suggestion
        count (int): The number of main blocks combined.
```

File: hamp_radar/iquick.py [hunk 4]
```diff
@@ -24,6 +55,25 @@ class SingleMainBlockGeometry:
...
+        step (Optional[int]): The distance between each SingleMainBlockGeometry
+                              instance, required if count > 1.
```
Comment (d70-t) [inline hamp_radar/iquick.py:74]:
```suggestion
        step (Optional[int]): The distance between each main block, required if count > 1.
```

File: hamp_radar/iquick.py [hunk 5]
```diff
@@ -43,12 +106,57 @@ def main_ofs(mainblock):
...
+    where the tag is the first 4 bytes of data and the pointer is the
+    next 4 bytes.
+
+    Args:
+        data (bytes): The data assumed to conform with Meteorological Ka-Band
+                      Cloud Radar MIRA35 Manual, section 2.3.1 'Definition of
+                      chunk common structure'
+
+    Returns:
+        Tuple[bytes, int]: The tag and the pointer for size.
```
Comment (d70-t) [inline hamp_radar/iquick.py:122]:
```suggestion
        Tuple[bytes, int]: The tag and the size.
```

File: hamp_radar/iquick.py [hunk 6]
```diff
@@ -43,12 +106,57 @@ def main_ofs(mainblock):
...
+    """
     return bytes(data[:4]), data[4:8].view("<i4")[0]
 
 
 def get_geometry(data):
+    """
+    Interprets and returns the geometry of the 'data' assuming it is a
+    memmap of a PDS file (or an open PDS file) for radar IQ data with structure
```
Comment (d70-t) [inline hamp_radar/iquick.py:130]:
```suggestion
    Interprets and returns the geometry of the 'data' assuming it is a "u1" `numpy`
    array with the contents of a PDS file for radar IQ data with structure
```

File: hamp_radar/iquick.py [hunk 7]
```diff
@@ -43,12 +106,57 @@ def main_ofs(mainblock):
...
+    as in Meteorological Ka-Band Cloud Radar MIRA35 Manual, section 2.3.2
+    'File structure'.
+
+    First attempt will read the main tag and size from first 8 bytes
+    of the data. If the data length is insufficient, or the main tag is not
+    found by attempting to return to the start of the file from main_size+8,
+    it shifts the offset by 1024 bytes (to skip file header) and tries again.
+    If the main tag is still not found, it raises a ValueError.
```
Comment (d70-t) [inline hamp_radar/iquick.py:138]:
```suggestion
    Tries to read the two consecutive main blocks at the beginning of the file,
    if this fails, assumes there's a 1024 byte header in the beginning and tries again.
    If the main tag is still not found, it raises a ValueError.
```

File: hamp_radar/iquick.py [hunk 8]
```diff
@@ -43,12 +106,57 @@ def main_ofs(mainblock):
...
+
+    If get_tag_size is sucessful, the function defines a generator
+    `main_blocks` that iterates through the data, yielding
+    `SingleMainBlockGeometry` instances for each block (i.e. for each main chunk
+    as defined in section 2.3.2)
```
Comment (d70-t) [inline hamp_radar/iquick.py:143]: This doesn't really describe the external behavior of the function, so I'd remove it from the docstring. The internal behavior can be seen when reading the code.
```suggestion
```

File: hamp_radar/iquick.py [hunk 9]
```diff
@@ -43,12 +106,57 @@ def main_ofs(mainblock):
...
+
+    The function returns a list of different MultiMainBlockGeometry instances
+    obtained from compacting together simliar SingleMainBlockGeometry from the
+    main_blocks multiple SingleMainBlockGeometry instances.
```
Comment (d70-t) [inline hamp_radar/iquick.py:147]: I'd also remove this one. The return value is described below and the way towards it is not relevant for external users.
```suggestion
```

File: hamp_radar/iquick.py [hunk 10]
```diff
@@ -43,12 +106,57 @@ def main_ofs(mainblock):
...
+
+    Parameters:
+    data (bytes): The binary data from which to extract geometry information.
```
Comment (d70-t) [inline hamp_radar/iquick.py:150]:
```suggestion
    data (ndarray, dtype "u1"): The binary data from which to extract geometry information.
```

File: hamp_radar/iquick.py [hunk 11]
```diff
@@ -73,6 +181,26 @@ def main_blocks(data, o):
 def compact_geometry(
     main_blocks: Iterable[SingleMainBlockGeometry],
 ) -> Iterable[MultiMainBlockGeometry]:
+    """
+    Compacts a sequence of SingleMainBlockGeometry instances into a sequence of
+    MultiMainBlockGeometry instances.
```
Comment (d70-t) [inline hamp_radar/iquick.py:186]: Sequences are much more specific than iterables. See also [Python ABC Collections](https://docs.python.org/3/library/collections.abc.html#collections-abstract-base-classes).
```suggestion
    Compacts an iterable of SingleMainBlockGeometry instances into an iterable of
    MultiMainBlockGeometry instances.
```

File: hamp_radar/iquick.py [hunk 12]
```diff
@@ -73,6 +181,26 @@ def main_blocks(data, o):
...
+
+    This function takes an iterable of SingleMainBlockGeometry objects and
+    combines sequential ones that are similar enough ('compatible') into a
+    single MultiMainBlockGeometry instance. When the previous
+    SingleMainBlockGeometry instance is not compatible to the current one, a new
+    MultiMainBlockGeometry instance is started. A list of all the 
+    MultiMainBlockGeometry instances is then returned.
```
Comment (d70-t) [inline hamp_radar/iquick.py:193]:
```suggestion
    MultiMainBlockGeometry instance is started. An iterable of all the 
    MultiMainBlockGeometry instances is then returned.
```

File: hamp_radar/iquick.py [hunk 13]
```diff
@@ -116,15 +244,41 @@ def compact_geometry(
 
 
 def extract_raw_arrays(data, mmbgs: Iterable[MultiMainBlockGeometry]):
+    """
+    Generator function to extract arrays from 'data' based on the
+    geometry given by the iterator over MultiMainBlockGeometry instances.
+
+    Iterating over a list of this generator yields a tuple for the subblocks
```
Comment (d70-t) [inline hamp_radar/iquick.py:251]:
```suggestion
    Iterating over a this generator yields a tuple for the subblocks
```

File: hamp_radar/iquick.py [hunk 14]
```diff
@@ -116,15 +244,41 @@ def compact_geometry(
...
+    across all the MultiMainBlockGeometry instances sequentially.
+
+    Optimisation uses NumPy library as_strided function to create a view of the
+    original data array interpreted with different shape and strides. Shape will
+    have (nrows, ncols) = (mmbg.count, block.size). Stride will be 1 unless
+    mmbg.count > 1, in which case mmbg.step is used to advance to the next
+    required subblock (skipping past other subblocks with different tags)
```
Comment (d70-t) [inline hamp_radar/iquick.py:258]: I'd also skip this, as this is an implementation detail and not super helpful for external users. Also I think especially with the use of keyword arguments, the code should now be pretty much equivalent to the text.
```suggestion
```

File: hamp_radar/iquick.py [hunk 15]
```diff
@@ -116,15 +244,41 @@ def compact_geometry(
...
+
+    Args:
+        data: The input data (memory map or open file) from which to extract the
+              arrays.
```
Comment (d70-t) [inline hamp_radar/iquick.py:262]:
```suggestion
        data: The input data (numpy array dtype "u1") from which to extract the arrays.
```

File: hamp_radar/iquick.py [hunk 16]
```diff
@@ -116,15 +244,41 @@ def compact_geometry(
...
+        mmbgs (Iterable[MultiMainBlockGeometry]): An iterable of
+              MultiMainBlockGeometry instances.
+
+    Yields:
+        tuple: A tuple containing:
+            - mmbg.tag: The tag/signature of the main block.
+            - block.tag: The tag/signature of the subblock.
+            - ndarray: A view of the data array corresponding to the subblock,
+                       created using numpy's as_strided function.
```
Comment (d70-t) [inline hamp_radar/iquick.py:271]: Again, it's an implementation detail, which probably doesn't help outside users.
```suggestion
            - ndarray: A view of the data array corresponding to the subblock.
```

File: hamp_radar/iquick.py [hunk 17]
```diff
@@ -186,28 +340,51 @@ def _decode(rawdata):
...
-    raw_arrays = list(extract_raw_arrays(data, get_geometry(data)))
+def read_pds(filename, postprocess=True):
+    """
+    Converts data from a file called 'filename', into an xarray Dataset. Currently
+    only functioning with geometry of pds files and decoders for IQ data of
+    Ka radar currently operational on HALO (last checked: 13th Septermber 2024).
+
+    Optimisation uses NumPy library memory-mapped array to avoid reading
+    entirety of large binary file into main memory from disc when only (small)
+    segment is desired.
```
Comment (d70-t) [inline hamp_radar/iquick.py:376]: I don't have a super strong opinion here, but this also is an implementation detail, not super relevant to external users.
```suggestion
```

File: hamp_radar/iquick.py [hunk 18]
```diff
@@ -222,26 +399,49 @@ def read_iq(filename, postprocess=True):
 
 
 def decode_time(ds):
+    """
+    Replaces 'Tm' and 'time_milli' variables in dataset 'ds' with decoded time.
+
+    The function calculates the time by adding:
+    - A base time of "1970-01-01"
+    - 'Tm' multiplied by 1 second (expressed as 1,000,000,000 ns)
+    - 'time_milli' multiplied by 0.001 (expressed as ns)
```
Comment (d70-t) [inline hamp_radar/iquick.py:408]: This text block is basically a verbatim copy of the code below, and also not helpful for outside users.
```suggestion
```

File: hamp_radar/iquick.py [hunk 19]
```diff
@@ -222,26 +399,49 @@ def read_iq(filename, postprocess=True):
...
+
+    The resulting time is then assigned to a new 'time' variable in the dataset.
```
Comment (d70-t) [inline hamp_radar/iquick.py:410]: Maybe this is redundant with the return value description.
```suggestion
```

File: hamp_radar/iquick.py [hunk 20]
```diff
@@ -222,26 +399,49 @@ def read_iq(filename, postprocess=True):
...
 def postprocess_iq(ds):
+    # TODO(ALL): move to new file
     return ds.pipe(decode_time)
 
 
 def main():
+    '''
+    Use an argument parser with positional argument for the filename to read IQ
+    data from the file called 'filename' and decode the time.
+    '''
```
Comment (d70-t) [inline hamp_radar/iquick.py:437]: I'd assume every `main()` function would do something like using an argument parse. The real docs for this function can be seen when running the script on the command line with `--help`.

```suggestion
```

**Review by d70-t — COMMENTED** [review]
Thanks for taking the effort of going through the code and writing all these docstrings. I believe they are indeed very helpful and should be included.
I've added some comments to the particular changes, which sometimes do slight wording changes. Sometimes I've also suggested to remove pieces, as they seemd to be redundant.

Possibly we might want to add some more type annotations in the longer run, which would probable reduce the amount of dosctrings needed even a bit more.

**Comment by yoctoyotta1024** [comment]
I completely agree with all your suggestions thanks Tobi! You're right alot of the notes were implentaton details (created just because I was making notes of what the code does for myself :)), we don't need those.

I'll just squash these commits and then I think this is ready to merge.

**Pull request merged**